sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,373 @@
+import torch
+from sbi.utils import gradient_ascent
+from sbi.utils.torchutils import ensure_theta_batched
+from sbi.inference.posteriors.base_posterior import NeuralPosterior
+
+# from sbi.utils.user_input_checks import process_x
+# from sbi.utils.sbiutils import match_theta_and_x_batch_shapes, within_support
+# from sbi.inference.potentials.base_potential import BasePotential
+
+from sbi.types import Shape
+from torch import Tensor
+from typing import List, Optional, Union, Callable
+
+
+class NeuralPosteriorEnsemble(NeuralPosterior):
+    r"""Wrapper class to bundle together different posterior instances into an ensemble.
+
+    This class creates a posterior ensemble from a set of N different, already trained
+    posterior estimators $p_{i}(\theta|x_o)$, where $i \in \{i,...,N\}$.
+
+    It can wrap all posterior classes available in sbi and even a mixture of different
+    posteriors, i.e. obtained via SNLE and SNPE at the same time, since it only
+    provides a pass-through to the class-methods of each posterior in the ensemble.
+
+    So far `log_prob()`, `sample()` and `map()` functionality are supported.
+
+    Example:
+
+    ```
+    import torch
+    from joblib import Parallel, delayed
+    from sbi.examples.minimal import simple
+
+    ensemble_size = 10
+    posteriors = Parallel(n_jobs=-1)(delayed(simple)() for i in range(ensemble_size))
+
+    ensemble = NeuralPosteriorEnsemble(posteriors)
+    ensemble.set_default_x(torch.zeros((3,)))
+    ensemble.sample((1,))
+    ```
+
+    Attributes:
```

Comment:
You can either remove this (we don't usually do it throughout the code-base) or add the missing attribute `potential_fn`